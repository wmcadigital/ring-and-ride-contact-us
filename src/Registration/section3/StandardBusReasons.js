import { useFormState } from "react-final-form";

import FormSection from "../../common/FormSection";
import ProgressIndicator from "../../common/ProgressIndicator";
import Question from "../../common/Question";
import getRequirementsSectionName from "./getRequirementsSectionName";
import CheckboxContainer from "../../common/CheckboxContainer";
import Checkbox from "../../common/Checkbox";
import TextArea from "../../common/TextArea";
import FieldError from "../../common/FieldError";
import { required } from "../../common/validation";

export const ReasonOptions = (registerForYourself) => {
  if (registerForYourself) {
    return {
      companionOrCarer: "I cannot travel without a companion or carer",
      limitedMobility: "I have limited mobility",
      longJourney: "The journey takes too long",
      notUnderstandService: "I don't understand how buses operate",
      fearAnxiety: "Fear and anxiety",
      cannotTravelBus: "I cannot travel on a bus",
    };
  }
  return {
    companionOrCarer: "They cannot travel without a companion or carer",
    limitedMobility: "They have limited mobility",
    longJourney: "The journey takes too long",
    notUnderstandService: "They don't understand how buses operate",
    fearAnxiety: "Fear and anxiety",
    cannotTravelBus: "They cannot travel on a bus",
  };
};

const StandardBusReasons = () => {
  const stateApi = useFormState();

  const formValues = stateApi.values;

  const registerForYourself = formValues["registerForYourself"] === "yes";
  const reasonOptions = ReasonOptions(registerForYourself);

  const question = "Which of the following best describes the reason for this?";

  const standardBusReasonError = stateApi.submitFailed
    ? stateApi.errors?.selectStandardBusReason
    : null;

  const otherReasonRequired = stateApi.submitFailed
    ? stateApi.errors?.standardBusReason?.otherReason
    : null;

  return (
    <FormSection>
      <ProgressIndicator
        sectionPosition="Section 3 of 3"
        sectionName={getRequirementsSectionName(stateApi)}
      />
      <Question text={question} />
      <p>Select all that apply</p>
      <CheckboxContainer error={standardBusReasonError}>
        <FieldError text={standardBusReasonError} />
        {Object.keys(reasonOptions).map((option) => (
          <Checkbox
            key={reasonOptions[option]}
            label={reasonOptions[option]}
            fieldName={`standardBusReason.${option}`}
          />
        ))}
        <Checkbox label="Other" fieldName="standardBusReason.other" />
        {formValues.standardBusReason?.other ? (
          <TextArea
            label="Please describe the reason"
            fieldName="standardBusReason.otherReason"
            validation={required}
            error={otherReasonRequired}
          />
        ) : null}
      </CheckboxContainer>
    </FormSection>
  );
};

export default StandardBusReasons;
