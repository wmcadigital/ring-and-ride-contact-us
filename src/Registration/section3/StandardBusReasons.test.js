import { useFormState } from "react-final-form";
import { createRenderer } from "react-test-renderer/shallow";

import StandardBusReasons from "./StandardBusReasons";

jest.mock("react-final-form");

describe("StandardBusReasons", () => {
  it("individual is filling out form for themselves", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<StandardBusReasons />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("individual is filling out form for someone else", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "no",
        firstName: "Test",
        lastName: "Testerson",
      },
    }));
    const renderer = createRenderer();

    renderer.render(<StandardBusReasons />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
      },
      submitFailed: true,
      errors: { selectStandardBusReason: "Select at least 1 topic" },
    }));
    const renderer = createRenderer();

    renderer.render(<StandardBusReasons />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an other reason text area if other reason selected", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
        standardBusReason: { other: true },
      },
    }));
    const renderer = createRenderer();

    renderer.render(<StandardBusReasons />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });

  it("renders with an error - other selected, no reason", () => {
    useFormState.mockImplementationOnce(() => ({
      values: {
        registerForYourself: "yes",
        standardBusReason: { other: true },
      },
      submitFailed: true,
      errors: { standardBusReason: { otherReason: "Required" } },
    }));
    const renderer = createRenderer();

    renderer.render(<StandardBusReasons />);

    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
});
