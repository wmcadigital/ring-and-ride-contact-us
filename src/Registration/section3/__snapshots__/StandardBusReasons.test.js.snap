// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardBusReasons individual is filling out form for someone else 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Applicant requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Which of the following best describes the reason for this?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error={null}
  >
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="standardBusReason.companionOrCarer"
      label="They cannot travel without a companion or carer"
    />
    <Checkbox
      fieldName="standardBusReason.limitedMobility"
      label="They have limited mobility"
    />
    <Checkbox
      fieldName="standardBusReason.longJourney"
      label="The journey takes too long"
    />
    <Checkbox
      fieldName="standardBusReason.notUnderstandService"
      label="They don't understand how buses operate"
    />
    <Checkbox
      fieldName="standardBusReason.fearAnxiety"
      label="Fear and anxiety"
    />
    <Checkbox
      fieldName="standardBusReason.cannotTravelBus"
      label="They cannot travel on a bus"
    />
    <Checkbox
      fieldName="standardBusReason.other"
      label="Other"
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`StandardBusReasons individual is filling out form for themselves 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Which of the following best describes the reason for this?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error={null}
  >
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="standardBusReason.companionOrCarer"
      label="I cannot travel without a companion or carer"
    />
    <Checkbox
      fieldName="standardBusReason.limitedMobility"
      label="I have limited mobility"
    />
    <Checkbox
      fieldName="standardBusReason.longJourney"
      label="The journey takes too long"
    />
    <Checkbox
      fieldName="standardBusReason.notUnderstandService"
      label="I don't understand how buses operate"
    />
    <Checkbox
      fieldName="standardBusReason.fearAnxiety"
      label="Fear and anxiety"
    />
    <Checkbox
      fieldName="standardBusReason.cannotTravelBus"
      label="I cannot travel on a bus"
    />
    <Checkbox
      fieldName="standardBusReason.other"
      label="Other"
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`StandardBusReasons renders with an error - other selected, no reason 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Which of the following best describes the reason for this?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer>
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="standardBusReason.companionOrCarer"
      label="I cannot travel without a companion or carer"
    />
    <Checkbox
      fieldName="standardBusReason.limitedMobility"
      label="I have limited mobility"
    />
    <Checkbox
      fieldName="standardBusReason.longJourney"
      label="The journey takes too long"
    />
    <Checkbox
      fieldName="standardBusReason.notUnderstandService"
      label="I don't understand how buses operate"
    />
    <Checkbox
      fieldName="standardBusReason.fearAnxiety"
      label="Fear and anxiety"
    />
    <Checkbox
      fieldName="standardBusReason.cannotTravelBus"
      label="I cannot travel on a bus"
    />
    <Checkbox
      fieldName="standardBusReason.other"
      label="Other"
    />
    <TextArea
      error="Required"
      fieldName="standardBusReason.otherReason"
      label="Please describe the reason"
      validation={[Function]}
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`StandardBusReasons renders with an error 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Which of the following best describes the reason for this?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error="Select at least 1 topic"
  >
    <FieldError
      text="Select at least 1 topic"
    />
    <Checkbox
      fieldName="standardBusReason.companionOrCarer"
      label="I cannot travel without a companion or carer"
    />
    <Checkbox
      fieldName="standardBusReason.limitedMobility"
      label="I have limited mobility"
    />
    <Checkbox
      fieldName="standardBusReason.longJourney"
      label="The journey takes too long"
    />
    <Checkbox
      fieldName="standardBusReason.notUnderstandService"
      label="I don't understand how buses operate"
    />
    <Checkbox
      fieldName="standardBusReason.fearAnxiety"
      label="Fear and anxiety"
    />
    <Checkbox
      fieldName="standardBusReason.cannotTravelBus"
      label="I cannot travel on a bus"
    />
    <Checkbox
      fieldName="standardBusReason.other"
      label="Other"
    />
  </CheckboxContainer>
</FormSection>
`;

exports[`StandardBusReasons renders with an other reason text area if other reason selected 1`] = `
<FormSection>
  <ProgressIndicator
    sectionName="Your requirements"
    sectionPosition="Section 3 of 3"
  />
  <Question
    text="Which of the following best describes the reason for this?"
  />
  <p>
    Select all that apply
  </p>
  <CheckboxContainer
    error={null}
  >
    <FieldError
      text={null}
    />
    <Checkbox
      fieldName="standardBusReason.companionOrCarer"
      label="I cannot travel without a companion or carer"
    />
    <Checkbox
      fieldName="standardBusReason.limitedMobility"
      label="I have limited mobility"
    />
    <Checkbox
      fieldName="standardBusReason.longJourney"
      label="The journey takes too long"
    />
    <Checkbox
      fieldName="standardBusReason.notUnderstandService"
      label="I don't understand how buses operate"
    />
    <Checkbox
      fieldName="standardBusReason.fearAnxiety"
      label="Fear and anxiety"
    />
    <Checkbox
      fieldName="standardBusReason.cannotTravelBus"
      label="I cannot travel on a bus"
    />
    <Checkbox
      fieldName="standardBusReason.other"
      label="Other"
    />
    <TextArea
      error={null}
      fieldName="standardBusReason.otherReason"
      label="Please describe the reason"
      validation={[Function]}
    />
  </CheckboxContainer>
</FormSection>
`;
